---
title: "Figure 3 Metzincin protease paper"
author: "Jasmin WÃ¤chter"
date: "2/9/2022"
output: html_document
---
```{r}

setwd("/Volumes/FREECOM HDD/R_ADAM")

library(ape)
library(DoubletFinder)
library(remotes)
library(Matrix)
library(fields)
library(KernSmooth)
library(ROCR)
library(parallel)
library(SingleCellExperiment)
library(Seurat)
library(tidyverse)
library(Matrix)
library(scales)
library(cowplot)
library(RCurl)
library(dplyr)
library(stringr)
library(ggplot2)
library(fields)
library(clustree)
library(ggraph)
library(plotly)
library(MAST)
library(multtest)
library(metap)

#library(monocle3)
library(monocle)
library(SeuratWrappers)
library(SCENT)
library(VGAM)
library(patchwork)
library(RColorBrewer)
library(plotly)


#Load the trophoblast object generated by Matthew Shannon
load("/Volumes/FREECOM HDD/R_ADAM/Trophoblast.Rdata")

```

#load previously identified metzincin
```{r}
Metzincin<-c("MMP2"  ,   "MMP3"  ,   "MMP11"  , "MMP14"  ,  "MMP15","MMP19",   "MMP23B",   "ADAMTS1"  ,"ADAMTS6" , "ADAMTS19" ,"ADAMTS20","PAPPA", "PAPPA2"  ,
"TIMP1"  ,  "TIMP2"   , "TIMP3"  ,  "RECK"  ,    "ADAM8"  ,  "ADAM9"  ,  "ADAM10"   ,"ADAM12" ,  "ADAM15" , 
"ADAM17" ,  "ADAM19"  , "ADAM28" ,  "TLL1"    , "BMP1"   ,  "TLL2" )

housekeep<-c("HLA-G","ERVFRD-1","TEAD4", "CDX2","NOTCH1")

```



#Monocle 2 
```{r}
gene_names<-Trophoblasts@assays$RNA@data@Dimnames[[1]]
gene_names<- HSMM@dispFitInfo$blind$disp_table$gene_id
```


#Prepare seurat object for Monocle!
```{r}

Trophoblasts@meta.data$cluster_name<- NA

 
Trophoblasts@meta.data$cluster_name[which(stringr::str_detect(Trophoblasts@meta.data$seurat_clusters, "^0"))] <- "CTB 1"
Trophoblasts@meta.data$cluster_name[which(stringr::str_detect(Trophoblasts@meta.data$seurat_clusters, "^1"))] <- "CTB 2"
Trophoblasts@meta.data$cluster_name[which(stringr::str_detect(Trophoblasts@meta.data$seurat_clusters, "^2"))] <- "CTB 3"
Trophoblasts@meta.data$cluster_name[which(stringr::str_detect(Trophoblasts@meta.data$seurat_clusters, "^3"))] <- "CTB 4"
Trophoblasts@meta.data$cluster_name[which(stringr::str_detect(Trophoblasts@meta.data$CellType, "^4"))] <- "cCTB 1"
Trophoblasts@meta.data$cluster_name[which(stringr::str_detect(Trophoblasts@meta.data$seurat_clusters, "^7"))] <- "cCTB 2"
Trophoblasts@meta.data$cluster_name[which(stringr::str_detect(Trophoblasts@meta.data$seurat_clusters, "^5"))] <- "EVT"
Trophoblasts@meta.data$cluster_name[which(stringr::str_detect(Trophoblasts@meta.data$seurat_clusters, "^6"))] <- "SCTp"

```



#A size factor relates to how many reads there are in each library. One can imagine that if you had two sample where 10% of the reads in each sample were from gene A, but in one sample 1M reads have been sequenced and in the other sample 2M reads had been sequenced then there would be a two fold increase in the counts from gene A in sample 2 compared to sample 1, but the actaul expression levels were the same.


#convert Seurat file into monocle 2 file 
```{r}



#Extract data, phenotype data, and feature data from the SeuratObject
data <- as(as.matrix(Trophoblasts@assays$RNA@data), 'sparseMatrix')

pd <- new('AnnotatedDataFrame', data = Trophoblasts@meta.data)

fData <- data.frame(gene_short_name = row.names(data), row.names = row.names(data))
fd <- new('AnnotatedDataFrame', data = fData)

#Construct monocle cds
monocle_cds <- newCellDataSet(data,
                         phenoData = pd,
                         featureData = fd,
                         lowerDetectionLimit = 0.5,
                         expressionFamily = negbinomial.size())
#I think with spars matrix is preferred 
HSMM <- newCellDataSet(
as(as.matrix(data), "sparseMatrix"),
phenoData = pd,
featureData = fd,
expressionFamily=negbinomial.size()
)
fData(HSMM)

#it is recommended to calculare size factors and dispersions before analysis

HSMM <- estimateSizeFactors(HSMM)
HSMM <- estimateDispersions(HSMM)

#If you are working with monocle2 you have to detach monocle3

HSMM <- detectGenes(HSMM, min_expr = 0.1)
print(head(fData(HSMM)))
```


```{r}
#SUBSET  genes 
expressed_genes <- row.names(subset(fData(HSMM),
    num_cells_expressed >= 10))

print(head(pData(HSMM)))

#look at mRNA distribution

pData(HSMM)$Total_mRNAs <- Matrix::colSums(exprs(HSMM))

HSMM <- HSMM[,pData(HSMM)$Total_mRNAs < 1e6]

upper_bound <- 10^(mean(log10(pData(HSMM)$Total_mRNAs)) +
            2*sd(log10(pData(HSMM)$Total_mRNAs)))
lower_bound <- 10^(mean(log10(pData(HSMM)$Total_mRNAs)) -
            2*sd(log10(pData(HSMM)$Total_mRNAs)))

qplot(Total_mRNAs, data = pData(HSMM),  geom =
"density") +
geom_vline(xintercept = lower_bound) +
geom_vline(xintercept = upper_bound)

#log transform

# Log-transform each value in the expression matrix.
L <- log(exprs(HSMM[expressed_genes,]))

# Standardize each gene, so that they are all on the same scale,
# Then melt the data with plyr so we can plot it easily
melted_dens_df <- reshape2::melt(Matrix::t(scale(Matrix::t(L))))

# Plot the distribution of the standardized gene expression values.
qplot(value, geom = "density", data = melted_dens_df) +
stat_function(fun = dnorm, size = 0.5, color = 'red') +
xlab("Standardized log(FPKM)") +
ylab("Density")

```
#Find correct clusters for Monocle 
```{r}

DefaultAssay  (object = Trophoblasts) <- "RNA"
EVT <- subset (Trophoblasts, subset =  
                 BCAM == 0 &
                 YAP1 == 0 &
                 TP63 == 0 &
                 TEAD4== 0 &
                 ITGA6== 0 &
                 GDF15==0&
               ITGA2==0)

DefaultAssay  (object = EVT) <- "integrated"
UMAPPlot (Trophoblasts,         label = TRUE) + NoLegend ()
UMAPPlot(EVT, label = TRUE) + NoLegend ()
```
#examine which cell clusters to assign
```{r}

DefaultAssay  (object = Trophoblasts) <- "RNA"
VlnPlot  (object = Trophoblasts, features = c("HLA-G"))
VlnPlot  (object = Trophoblasts, features = c("ITGA5"))
VlnPlot  (object = Trophoblasts, features = c("EGFR"))
VlnPlot  (object = Trophoblasts, features = c("BCAM"))
VlnPlot  (object = Trophoblasts, features = c("NOTCH2"))
VlnPlot  (object = Trophoblasts, features = c("ITGA2"))
VlnPlot  (object = Trophoblasts, features = c("TP63"))
VlnPlot  (object = Trophoblasts, features = c("TEAD4"))
VlnPlot  (object = Trophoblasts, features = c("ITGA6"))
VlnPlot  (object = Trophoblasts, features = c("ERVV-1"))
VlnPlot  (object = Trophoblasts, features = c("ERVFRD-1"))
VlnPlot  (object = Trophoblasts, features = c("BMP7"))
VlnPlot  (object = Trophoblasts, features = c("BMP4"))
VlnPlot  (object = Trophoblasts, features = c("CCNA2"))
VlnPlot  (object = Trophoblasts, features = c("PCNA"))
VlnPlot  (object = Trophoblasts, features = c("MKI67"))
VlnPlot  (object = Trophoblasts, features = c("CGA"))
VlnPlot  (object = Trophoblasts, features = c("GDF15"))
VlnPlot  (object = Trophoblasts, features = c("CDH1"))
VlnPlot  (object = Trophoblasts, features = c("ELF5"))
VlnPlot  (object = Trophoblasts, features = c("FBN1"))


```



#assign cell type clusters
```{r}
HLA_G_id <- row.names(subset(fData(HSMM), gene_short_name == "HLA-G"))

ITGA5_id <- row.names(subset(fData(HSMM), gene_short_name == "ITGA5"))
EGFR_id <- row.names(subset(fData(HSMM),
    gene_short_name == "EGFR"))
BCAM_id <- row.names(subset(fData(HSMM), gene_short_name == "BCAM"))

ERVV1_id <- row.names(subset(fData(HSMM),
    gene_short_name == "ERVV-1"))
ERVFRD1_id <- row.names(subset(fData(HSMM), gene_short_name == "ERVFRD-1"))

CGA_id <- row.names(subset(fData(HSMM),
   gene_short_name == "CGA"))
GDF15_id <- row.names(subset(fData(HSMM),
   gene_short_name == "GDF15"))

ELF5_id <- row.names(subset(fData(HSMM),
   gene_short_name == "ELF5"))

cth <- newCellTypeHierarchy()


cth <- addCellType(cth, "EVT", classify_func =
                     function(x) 
                       { x[HLA_G_id,] >= 1.5 |  x[ITGA5_id,] >= 2})



cth <- addCellType(cth, "CTB", classify_func =
                     function(x) 
                       {x[BCAM_id,]> 1.5  |x[EGFR_id,]> 2 })

cth <- addCellType(cth, "SCT", classify_func =
                     function(x) 
                       { x[ERVFRD1_id,] >= 0.5 |  x[ERVV1_id,] >= 1 |  x[CGA_id,] >= 6 |  x[GDF15_id,] >= 4.5})


HSMM <- classifyCells(HSMM,cth,0.1)
table(pData(HSMM)$CellType)

pie <- ggplot(pData(HSMM),
aes(x = factor(1), fill = factor(CellType))) + geom_bar(width = 1)
pie + coord_polar(theta = "y") +
theme(axis.title.x = element_blank(), axis.title.y = element_blank())

```
#Clustering using marker genes 
```{r}
marker_diff <- markerDiffTable(HSMM[expressed_genes,],
            cth,
            cores = 1)

candidate_clustering_genes <-
    row.names(subset(marker_diff, qval < 0.01))
marker_spec <-
  calculateMarkerSpecificity(HSMM[candidate_clustering_genes,], cth)
head(selectTopMarkers(marker_spec, 6))

 
semisup_clustering_genes <- unique(selectTopMarkers(marker_spec, 500)$gene_id)
HSMM <- setOrderingFilter(HSMM, semisup_clustering_genes)
plot_ordering_genes(HSMM)
```

```{r}
plot_pc_variance_explained(HSMM, return_all = F)
```

```{r}
HSMM <- reduceDimension(HSMM, max_components = 2, num_dim = 3,
  norm_method = 'log',
  reduction_method = 'tSNE',
  residualModelFormulaStr = "~GA",
  verbose = T)
```

```{r}
HSMM <- clusterCells(HSMM, num_clusters = 6)
plot_cell_clusters(HSMM, 1, 2, color = "CellType")

```
#Trajectory Step1
```{r}

HSMM <- setOrderingFilter(HSMM, semisup_clustering_genes)
plot_ordering_genes(HSMM)
```
#STep2: reduce data dimensionality
```{r}
HSMM <- reduceDimension(HSMM, max_components = 2,
    method = 'DDRTree')
```
#Step3: order cells along the trajectory
```{r}
#first order then root state... 
HSMM<-orderCells(HSMM)
```

#fixing monocle names
```{r}
HSMM$cluster_name<- NA
 
HSMM$cluster_name[which(stringr::str_detect(HSMM$seurat_clusters, "^0"))] <- "CTB 1"
HSMM$cluster_name[which(stringr::str_detect(HSMM$seurat_clusters, "^1"))] <- "CTB 2"
HSMM$cluster_name[which(stringr::str_detect(HSMM$seurat_clusters, "^2"))] <- "CTB 3"
HSMM$cluster_name[which(stringr::str_detect(HSMM$seurat_clusters, "^3"))] <- "CTB 4"
HSMM$cluster_name[which(stringr::str_detect(HSMM$seurat_clusters, "^4"))] <- "cCTB 1"
HSMM$cluster_name[which(stringr::str_detect(HSMM$seurat_clusters, "^5"))] <- "EVT"
HSMM$cluster_name[which(stringr::str_detect(HSMM$seurat_clusters, "^6"))] <- "SCTp"
HSMM$cluster_name[which(stringr::str_detect(HSMM$seurat_clusters, "^7"))] <- "cCTB 2"


```

```{r}

plot_cell_trajectory(HSMM, color_by = "Pseudotime")+theme(legend.position =c(.75,.85), legend.direction = "horizontal")
```


```{r}  


plot_cell_trajectory(HSMM, color_by = "seurat_clusters",
                      cols       = c('0' = '#CD9600',
                                     "1"= "#7CAE00",
                                     "2" = "#00BE67",
                                     '3' = '#00BFC4',
                                     '4'    = '#00A9FF',
                                     '5'   = '#FF61CC',
                                     '6'   = '#F8766D',
                                     "7" = '#C77CFF'))+ scale_color_manual(breaks = c("0", "1", "2",'3','4' ,'5', "6","7"), values=c('#CD9600',"#7CAE00","#00BE67",'#00BFC4','#00A9FF','#FF61CC','#F8766D', '#C77CFF'))





plot_cell_trajectory(HSMM, color_by = "seurat_clusters")
plot_cell_trajectory(HSMM, color_by = "CellType")
```

```{r}
GM_state <- function(cds){
  if (length(unique(pData(cds)$State)) > 1){
    T0_counts <- table(pData(cds)$State, pData(cds)$seurat_clusters)[,"1"]
    return(as.numeric(names(T0_counts)[which
          (T0_counts == max(T0_counts))]))
  } else {
    return (1)
  }
}

HSMM<- orderCells(HSMM , root_state = GM_state(HSMM))
```

#Check ordering
```{r}

plot_cell_trajectory(HSMM, color_by = "Pseudotime")+theme(legend.position =c(.75,.85), legend.direction = "horizontal")
  


plot_cell_trajectory(HSMM, color_by = "cluster_name",
                      cols       = c('cCTB 1' = '#CD9600',
                            'cCTB 2'   = '#C77CFF',
                            'CTB 1'  = '#F8766D',
                            'CTB 2'  = '#0CB702',
                            'CTB 3' = '#00BFC4',
                            'EVT'    = '#00A9FF',
                            'SCTp'   = '#FF61CC'))+ scale_color_manual(breaks = c('cCTB 1',"cCTB 2", "CTB 1", "CTB 2",'CTB 3','EVT' ,'SCTp'), values=c('#CD9600','#C77CFF', '#F8766D', '#0CB702',"#00BFC4",'#00A9FF','#FF61CC'))




plot_cell_trajectory(HSMM, color_by = "Pseudotime")
```

#Fig 2A
```{r}

jpeg("Monocle2_pseudotime.jpeg", units="in", width=6, height=5, res=300)

plot_cell_trajectory(HSMM, color_by = "seurat_clusters",
                      cols       = c('0' = '#CD9600',
                                     "1"= "#7CAE00",
                                     "2" = "#00BE67",
                                     '3' = '#00BFC4',
                                     '4'    = '#00A9FF',
                                     '5'   = '#FF61CC',
                                     '6'   = '#F8766D',
                                     "7" = '#C77CFF'))+ scale_color_manual(breaks = c("0", "1", "2",'3','4' ,'5', "6","7"), values=c('#CD9600',"#7CAE00","#00BE67",'#00BFC4','#00A9FF','#FF61CC','#F8766D', '#C77CFF'))

dev.off()
```


#Figure 2B metzincin
```{r}
#BEAM produces fancy heatmaps and fancy pseudotime plots that are bidirectional


BEAM_res1 <- BEAM(HSMM[c(Metzincin),], branch_point = 1, cores = 1)
BEAM_res1 <- BEAM_res2[order(BEAM_res1$qval),]
BEAM_res1 <- BEAM_res2[,c("gene_short_name", "pval", "qval")]



jpeg("Metzincin_pseudotime_NOhousekeep.jpeg", units="in", width=6, height=5, res=300)
plot_genes_branched_heatmap(HSMM[row.names(subset(BEAM_res1,
                                          qval < 1e-4)),],
                                          branch_point = 1,
                                          num_clusters = 1,
                                          cores = 1,
                            branch_labels = c("EVT","SCT"),
                                          use_gene_short_name = T,
                                          show_rownames = T)

dev.off()
```

#Figure 2B housekeep
```{r}
#BEAM produces fancy heatmaps and fancy pseudotime plots that are bidirectional



BEAM_res4 <- BEAM(HSMM[c(housekeep),], branch_point = 1, cores = 1)
BEAM_res4 <- BEAM_res4[order(BEAM_res4$qval),]
BEAM_res4 <- BEAM_res4[,c("gene_short_name", "pval", "qval")]

View(BEAM_res4)

jpeg("Metzincin_pseudotime_housekeep.jpeg", units="in", width=6, height=5, res=300)
plot_genes_branched_heatmap(HSMM[row.names(subset(BEAM_res4,
                                          qval < 1e-4)),],
                                          branch_point = 1,
                                          num_clusters = 2,
                                          cores = 1,
                            branch_labels = c("EVT","SCT"),
                                          use_gene_short_name = T,
                                          show_rownames = T)
dev.off()

```



